File: data_structures/go/linked_list_test.go
package ds

import (
	"testing"

	"gotest.tools/assert"
)

func TestLinkedListAdd(t *testing.T) {
	t.Log("Testing that we can create LinkedLists and add values to it.")
	ll := LinkedList{}
	assert.Equal(t, 0, ll.Size())

	ll.AddHead(1)
	ll.AddHead(2)
	ll.AddHead(3)

	assert.Equal(t, 3, ll.Size())

	ll.AddTail(4)
	ll.AddTail(5)
	ll.AddTail(6)

	assert.Equal(t, 6, ll.Size())
}

func TestLinkedListPop(t *testing.T) {
	t.Log("Testing that popping values from the list gives what we expect.")
	ll := LinkedList{}

	ll.AddTail(4)
	ll.AddTail(5)
	ll.AddTail(6)

	assert.Equal(t, 4, ll.PopHead())
	assert.Equal(t, 6, ll.PopTail())
	assert.Equal(t, 5, ll.PopTail())
	assert.Equal(t, 0, ll.Size())
}

func TestLinkedListPeek(t *testing.T) {
	ll := LinkedList{}

	ll.AddHead(9)
	assert.Equal(t, 9, ll.PeekHead())
	assert.Equal(t, 9, ll.PeekTail())

	ll.AddTail(5)
	assert.Equal(t, 9, ll.PeekHead())
	assert.Equal(t, 5, ll.PeekTail())
	assert.Equal(t, 2, ll.Size())
}

func TestLinkedListDisplay(t *testing.T) {

	t.Log("Testing that we can create LinkedLists and add values to it.")
	ll := LinkedList{}

	ll.AddHead(1)
	ll.AddHead(3)
	ll.AddHead(5)
	ll.AddTail(2)
	ll.AddTail(4)
	ll.AddTail(6)
	ll.Display()

}

File: data_structures/cpp/linked_list_test.cpp
#include <bits/stdc++.h>
#define CATCH_CONFIG_MAIN  // This tells Catch to provide a main() - only do this in one cpp file
#include "catch.hpp"
#include "linked_list.hpp"

using namespace std;

TEST_CASE("Can add items to LL.", "[LinkedList]") {
    LinkedList<int> ll;
    ll.addHead(1);
    ll.addHead(3);
    ll.addHead(5);
    REQUIRE(ll.size() == 3);

    ll.addTail(2);
    ll.addTail(4);
    ll.addTail(6);
    REQUIRE(ll.size() == 6);
}

TEST_CASE("Can pop items from LL.", "[LinkedList]") {
    LinkedList<int> ll;
    ll.addTail(1);
    ll.addTail(3);
    ll.addTail(5);

    REQUIRE(ll.popHead() == 1);
    REQUIRE(ll.popTail() == 5);
    REQUIRE(ll.popHead() == 3);
    REQUIRE(ll.size() == 0);
}

TEST_CASE("We can display items.", "[LinkedList]") {
    LinkedList<int> ll;
    ll.addTail(1);
    ll.addTail(3);
    ll.addTail(5);
    ll.addHead(4);
    ll.addHead(2);
    ll.addHead(6);
    ll.display();
}